src/censored.jl

```diff
@@ -0,0 +1,409 @@
+    return censored(
+        d.uncensored,
+        d.lower === missing ? l : max(l, d.lower),
+        d.upper === missing ? u : min(u, d.upper),
+    )
+end
+function censored(d::Censored, ::Missing, u::Real)
+    return censored(d.uncensored, d.lower, d.upper === missing ? u : min(u, d.upper))
+end
+function censored(d::Censored, l::Real, ::Missing)
+    return censored(d.uncensored, d.lower === missing ? l : max(l, d.lower), d.upper)
+end
+
+function params(d::Censored)
+    d0params = params(d.uncensored)
+    return if d.lower === missing
+        (d0params..., d.upper)
+    elseif d.upper === missing
+        (d0params..., d.lower)
+    else
+        (d0params..., d.lower, d.upper)
+    end
+end
+
+partype(d::Censored) = partype(d.uncensored)
+
+Base.eltype(::Type{<:Censored{D,S,T}}) where {D,S,T} = promote_type(T, eltype(D))
+
+#### Range and Support
+
+function islowerbounded(d::Censored)
+    return (
+        islowerbounded(d.uncensored) ||
+        (d.lower !== missing && cdf(d.uncensored, d.lower) > 0)
+    )
+end
+function isupperbounded(d::Censored)
+    return (
+        isupperbounded(d.uncensored) ||
+        (d.upper !== missing && _ccdf_inc(d.uncensored, d.upper) > 0)
+    )
+end
+
+function minimum(d::Censored)
+    d0min = minimum(d.uncensored)
+    return d.lower === missing ? min(d0min, d.upper) : max(d0min, d.lower)
+end
+
+function maximum(d::Censored)
+    d0max = maximum(d.uncensored)
+    return d.upper === missing ? max(d0max, d.lower) : min(d0max, d.upper)
+end
+
+function insupport(d::Censored{<:UnivariateDistribution}, x::Real)
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    return (
+        _in_open_interval(x, lower, upper) ||
+        (_eqnotmissing(x, lower) && cdf(d0, lower) > 0) ||
+        (_eqnotmissing(x, upper) && _ccdf_inc(d0, upper) > 0)
+    )
+end
+
+#### Show
+
+function show(io::IO, d::Censored)
+    print(io, "Censored(")
+    d0 = d.uncensored
+    uml, namevals = _use_multline_show(d0)
+    uml ? show_multline(io, d0, namevals; newline=false) : show_oneline(io, d0, namevals)
+    print(io, ", range=(", d.lower, ", ", d.upper, "))")
+end
+
+_use_multline_show(d::Censored) = _use_multline_show(d.uncensored)
+
+
+#### Statistics
+
+quantile(d::Censored, p::Real) = _clamp(quantile(d.uncensored, p), d.lower, d.upper)
+
+median(d::Censored) = _clamp(median(d.uncensored), d.lower, d.upper)
+
+# the expectations use the following relation:
+# 𝔼_{x ~ d}[h(x)] = P_{x ~ d₀}(x < l) h(l) + P_{x ~ d₀}(x > u) h(u)
+#                 + P_{x ~ d₀}(l ≤ x ≤ u) 𝔼_{x ~ τ}[h(x)],
+# where d₀ is the uncensored distribution, d is d₀ censored to [l, u],
+# and τ is d₀ truncated to [l, u]
+
+function mean(d::Censored)
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    prob_lower = lower === missing ? 0 : _cdf_noninc(d0, lower)
+    prob_upper = upper === missing ? 0 : ccdf(d0, upper)
+    prob_interval = 1 - (prob_lower + prob_upper)
```

Comment:
Should we work with log probabilities here?

Reply:
I can do that. I suspect it won't help much in the end, since `Truncated` itself uses direct probabilities.

Reply:
We use log probabilities in the construction of `Truncated`: https://github.com/JuliaStats/Distributions.jl/blob/ad2c3bc245bdfaf633164b996f2e9043da35e3bb/src/truncate.jl#L23-L37 I'm pretty sure there was a bug due to an earlier version not being based on log probabilities.

Reply:
What I mean is it eventually exponentiates that logarithm and then re`-log`s it when necessary, which seems like it will cause problems:
https://github.com/JuliaStats/Distributions.jl/blob/7206033e2113ded8758bcaa408deadb5156500ec/src/truncate.jl#L106

e.g.
```julia
julia> logx = -1e4
-10000.0

julia> logx - log(exp(logx))
Inf
```

Reply:
See the latest commit. Working on log scale proves to be problematic for testing censoring to bounded intervals containing no probability (i.e. `prob_interval = 0`, essentially the only example we can test is a `MixtureModel` with a gap in the middle containing no support), because if we work on log scale, then `log_prob_interval` might end up being a small number, not exactly 0, and if we miss that check and proceed to truncation, an error is raised because for `MixtureModel`, the expectations are not defined. It's not a huge problem, but I need to think of a way to test that branch.